### PR TITLE
Custom serializer support - Expose log in Request

### PR DESCRIPTION
**Pull request by delvedor**
- Now the logger instance is exposed in the Request object, `req.req.log.fn()` was a little bit unintuitive.

- `reply.send()` now accepts two parameters, the payload and a conditional boolean, that disables the custom serializer and uses the default one.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -232,12 +232,13 @@ An object including the following properties:
...
 * `params` - the params matching the URL
 * `req` - the incoming HTTP request from Node core
+* `log` - the logger instance of the incoming request
 
 <a name="reply"></a>
 #### Reply
 
 An object that exposes three APIs.
-* `.send(payload)` - Sends the payload to the user, could be a plain text, JSON, stream, or an Error object.
+* `.send(payload, [serializeSchema])` - Sends the payload to the user, could be a plain text, JSON, stream, or an Error object. *serializeSchema* if setted to `false`, disables the custom serializer for the current handler and uses the default one.
```
Comment (mcollina) [inline README.md]: I'm 👎  on this change, as I think overloaded methods generally adds complexity to Node.js code.
How about `reply.setSerializer(null)` or `reply.serialize(null)`? Then we can also support custom serializers, e.g. [msgpack5](https://github.com/mcollina/msgpack5)

Comment (delvedor) [inline README.md]: Love the idea, is definitely a better solution (and more flexible)!

File: lib/reply.js [hunk 2]
```diff
@@ -76,6 +84,11 @@ Reply.prototype.code = function (code) {
   return this
 }
 
+Reply.prototype.serializer = function (fn) {
+  this._serializer = fn
+  return this
+}
```
Comment (mcollina) [inline lib/reply.js]: should we allow to set a custom content-type straight from this function? it seems a must-have.

File: package.json [hunk 3]
```diff
@@ -59,13 +59,13 @@
...
-    "fast-safe-stringify": "^1.1.6",
+    "fast-json-stringify": "^0.10.4",
+    "fast-safe-stringify": "^1.1.11",
     "fastseries": "^1.7.2",
     "middie": "^0.1.1",
     "pathname-match": "^1.2.0",
     "pino-http": "^2.4.2",
     "pump": "^1.0.2",
-    "wayfarer": "^6.4.1"
+    "wayfarer": "^6.5.0"
```
Comment (mcollina) [inline package.json]: do you update those manually all the time, or you use some script? It's not needed, the new versions are picked by the semver range.

Comment (delvedor) [inline package.json]: I use [ncu](https://www.npmjs.com/package/npm-check-updates) to check updates, and if there are some I use to update and run all the test, to find out if there some issue.
Shall we introduce a tool like [greenkeeper](https://greenkeeper.io/)?

File: lib/reply.js (hunk 2 above)
Comment (delvedor) [inline lib/reply.js]: As you can see [here](https://github.com/fastify/fastify/blob/6adad7d91684249d5f208b12fb6e0b33b20f6c96/lib/reply.js#L58) and [here](https://github.com/fastify/fastify/blob/6adad7d91684249d5f208b12fb6e0b33b20f6c96/lib/reply.js#L67), we are assuming that the user knows which `content-type` he is using.
Probably we must add this to the docs.

Comment (mcollina) [inline lib/reply.js]: Ahum, this custom serializer logic might require some more thinking.
We should let devs sets a serializer both at the `Reply`  level, but also at the route level.

For this change, I mean:

```js
reply.serializer('application-json', (res) => `"this is JSON"`)
```

👍  for the docs

File: package.json (hunk 3 above)
Comment (mcollina) [inline package.json]: ncu is ok.  Are you using it with --upgradeAll? In theory all those changes should not be happening: https://www.npmjs.com/package/npm-check-updates#why-is-it-not-updating-100-to-101-when-101-is-the-latest.

I'm ok with greenkeeper.

File: lib/reply.js (hunk 2 above)
Comment (delvedor) [inline lib/reply.js]: I'm rewriting the docs, in the next week will land in pr.
If you are ok we can merge this as it is, and go deeper with this argument in the new docs.

**Pull request merged**